### PR TITLE
Add HER to DDPG

### PR DESCRIPTION
**Pull request by CatherineSue**
Add Hindsight Experience Replay(HER) into TF. This includes:
 - Implement a HER replay buffer into garage. Also rewrite the regular
   replay buffer to make the two buffers has consistent interfaces.
 - Add a sampling strategy of HER replay buffer. Currently we use the
   future strategy in the original paper.
 - Add HER into DDPG. This part basically breaks with original DDPG
   which contains some repetitive code. Will fix later.
 - Add a benchmark test for HER + DDPG.

Fixes: #209

### CODE REVIEW

**Comment by CatherineSue** [comment]
# Benchmarks 
This is the regression test result of HerDdpg benchmark over garage HER_DDPG and baselines HER_DDPG; each plot represents the SuccessRate over epochs in a particular task. 

## Hyperparameters
```
"actor_lr": 1e-3,
"critic_lr": 1e-3,
"actor_hidden_sizes": [256, 256, 256],
"critic_hidden_sizes": [256, 256, 256],
"n_epochs": 50,
"n_epoch_cycles": 20,
"n_rollout_steps": 100,
"n_train_steps": 40,
"discount": 0.9,
"tau": 0.05,
"replay_buffer_size": int(1e6),
"sigma": 0.2,
"batch_size": 256,
```

## FetchReach-v1
fetchreach-v1_benchmark

File: garage/replay_buffer/her_replay_buffer.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                                      *transitions[k].shape[1:])
+            for k in transitions.keys()
+        }
+        assert (transitions["action"].shape[0] == sample_batch_size)
+        return transitions
+
+    return _her_sample_transitions
+
+
+class HerReplayBuffer(object):
```
Comment (ryanjulian) [inline garage/replay_buffer/her_replay_buffer.py]: Python 3 classes don't need to inherit from object:

`class HerReplayBuffer:`

File: garage/replay_buffer/her_replay_buffer.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            for key in self._episode_buffer.keys():
+                self._episode_buffer[key].clear()
+
+    @property
+    def full(self):
+        """Whether the buffer is full."""
+        return self._current_size == self._size
+
+    def _get_storage_idx(self, size_increment=None):
+        """Get the storage index for the episode to add into the buffer."""
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply garage/replay_buffer/her_replay_buffer.py]: Yes.

File: garage/replay_buffer/her_replay_buffer.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+"""
+import inspect
+from typing import Callable
+
+import numpy as np
+
+from garage.misc.overrides import overrides
+
+
+def make_her_sample(replay_k: int, reward_fun: Callable):
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply garage/replay_buffer/her_replay_buffer.py]: Done.

File: garage/replay_buffer/her_replay_buffer.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        self._episode_buffer = {}
+        self._size = size_in_transitions // time_horizon
+        for key in buffer_shapes.keys():
+            self._episode_buffer[key] = list()
+        self._buffer = {
+            key: np.empty([self._size, *shape])
+            for key, shape in buffer_shapes.items()
+        }
+
+    @overrides
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply garage/replay_buffer/her_replay_buffer.py]: Done.

File: garage/tf/algos/ddpg.py [hunk 5]
```diff
@@ -56,6 +59,12 @@ def __init__(self,
                  pause_for_plot=False,
                  actor_optimizer=None,
                  critic_optimizer=None,
+                 use_her=False,
+                 clip_obs=200.,
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply garage/tf/algos/ddpg.py]: Done.

File: garage/replay_buffer/her_replay_buffer.py (hunk 1 above)
Comment (CatherineSue) [reply garage/replay_buffer/her_replay_buffer.py]: Done.

Comment (CatherineSue) [reply garage/replay_buffer/her_replay_buffer.py]: Done.

File: garage/replay_buffer/her_replay_buffer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhanpenghe) [reply garage/replay_buffer/her_replay_buffer.py]: If we don't need it, yes we can just keep it simple.

**Comment by CatherineSue** [comment]
No. But since I changed the ReplayBuffer interface, it breaks theano/algos/ddpg. That's why the CI fails. Fixing it.

**Comment by ryanjulian** [comment]
Thanks! Let's not submi anything broken.

On Fri, Aug 10, 2018, 10:25 AM Chang Su <notifications@github.com> wrote:

> No. But since I changed the ReplayBuffer interface, it breaks
> theano/algos/ddpg. That's why the CI fails. Fixing it.
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_rlworkgroup_garage_pull_265-23issuecomment-2D412150159&d=DwMCaQ&c=clK7kQUTWtAVEOVIgvi0NU5BOUHhpN0H8p7CSfnc_gI&r=VEuK-fS6-fUHjpi4P-BnNQ&m=u_1cGb56rOxTea82CQ876IUS3THK1teK7FlhBOvpsCA&s=_uGA6cve0rlmZ_kIazCSWQAgD_u350bCDu1Kors8lUM&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABT-5F2-2DKF-2DDrIrDo-5Fpqx9-2DPjEEqhrUoMjks5uPcHzgaJpZM4V3Z89&d=DwMCaQ&c=clK7kQUTWtAVEOVIgvi0NU5BOUHhpN0H8p7CSfnc_gI&r=VEuK-fS6-fUHjpi4P-BnNQ&m=u_1cGb56rOxTea82CQ876IUS3THK1teK7FlhBOvpsCA&s=R1Asu8H6TtpUUj1WLXkj7kHDn1JSgvswN7QguaSwOqo&e=>
> .
>
-- 
PhD Student
USC CLMC/RESL
(510) 859-3224
rjulian@usc.edu

**Pull request merged**